### PR TITLE
Process Google Calendar-private extended property "iCalUid"

### PR DESCRIPTION
**Pull request by ArnyminerZ**
Depends on https://github.com/bitfireAT/davx5-ose/issues/538
Depends on #128
Depends on #130

### CODE REVIEW

**Comment by ArnyminerZ** [comment]
Note: I've just added the tests, I've not checked whether it works with Google Calendar.

File: lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
Comment (rfc2822) [inline lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt:305]: Currently UIDs are handled by the calling app (DAVx5).

**Review by rfc2822 — CHANGES_REQUESTED** [review]
Modify UID handling outside of the narrow scope of the issue (adding `iCalUid` support) could break many things.

Currently the calling app (DAVx5) handles UIDs in a custom way. We have to synchronize UID usage between DAVx5 and ical4android first. Then we can extend ical4android to provide other methods to provide the UID.

I have added a dependency for the issue.

**Comment by ArnyminerZ** [comment]
Everything should be ready here then @rfc2822. If you prefer to test the previous change before, we can wait to merge.

File: lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
Comment (rfc2822) [inline lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt:918]: Is this a mistake?

**Review by rfc2822 — COMMENTED** [review]
Can you please rebase/merge to the new integrated UID handling

File: lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
Comment (ArnyminerZ) [reply lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt:919]: Yup, my bad. Fixed

File: lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
Comment (rfc2822) [inline lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt:87]: We should decide how to name the constants that contain `ExtendedProperties.NAME`s and then apply the same scheme to all properties – otherwise confusion may occur at some point.

I originally used the `MIMETYPE_` prefix because I think they should be formatted as MIME types (found that somewhere in the docs). However in reality they're often just a string (like `iCalUid`).

So we should either name everything `MIMETYPE_`, then it would be `MIMETYPE_ICAL_UID`.

Or (maybe better) we should prefix everything (including `CATEGORIES` and `URL`) with `EXTNAME_`. But in any case all things of the same type should have the same prefix (and defined alphabetically so that a certain property can be found quickly).

File: lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
Comment (rfc2822) [inline lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt:306]: I think this duplicates the `event.uid = row.getAsString(Events.UID_2445)` a few lines above.

File: lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
Comment (rfc2822) [inline lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt:1665]: Not from this PR, but it's named "text…" instead of "test…"

File: lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
Comment (rfc2822) [inline lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt:1458]: We can set this to `= {}` for the case that we only want to test only an extended property (and remove all `valuesBuilder = {}`).

**Review by rfc2822 — COMMENTED** [review]
Basically looks good! Only a few comments.

File: lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
Comment (ArnyminerZ) [reply lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt:87]: I prefer the `EXTNAME_` syntax better than `MIMETYPE_`.

**Review by rfc2822 — COMMENTED** [review]
Added a (now failing) test for `UID_2445` vs `iCalUid` precedence.

I think `UID_2445` should have precedence over `iCalUid` when both values are present (like for instance many local Google Calendar-imported events now which were set to an empty `UID_2445` and thus got a new UUID from DAVx5 while having another `iCalUid`) because `UID_2445` is the standard property and `iCalUid` is only a proprietary extension that shall be supported to optimize processing Google Calendar events. In doubt, `UID_2445` is the more important property.

This also leads to the next thing that has to be clarified: how do we deal with `iCalUid` when creating / updating events? Currently we don't create it (which is good I think), however we also don't update / delete it when the `uid` is changed.

So the question is whether we should:

- Update the property in `AndroidEvent.update()` (which also updates the `UID_2445`) and/or
- Delete `EXTNAME_ICAL_UID` rows in `AndroidEvent.update()` (like we delete the other supported `EXTNAME_` properties; however they will be re-created by the corresponding `insert...` Methods) and/or
- Keep it like it is? But then the iCalUid can be different from `UID_2445` although we have written `AndroidEvent` to the calendar provider, which should put the event into a well-known state. The question is what this means for us and whether it would have potential negative consequences.

Currently I'm favoring option 2 (delete the row). Then we would support `iCalUid` when reading; but when writing, we would make sure that only `UID_2445` is used.

**Comment by ArnyminerZ** [comment]
I'd just remove the `iCalUid` property and replace it with `UID_2445`, so `iCalUid` is read-only, and we actually use `UID_2445` when updating. AFAIK Google Calendar, which is the one adding this custom property, shouldn't have any problems with `UID_2445`, right? So I'd just get rid of it ASAP.

**Comment by ArnyminerZ** [comment]
Well, I don't love the approach, but this is the simplest I've been able to achieve. I've split `populateEvent` into `populateAndroidEvent` because I needed direct access to `AndroidEvent` to test the `extendedProperties` variable. If you feel like it's not necessary to test that it's initialized correctly (since we are already kind of testing it in `testPopulateEvent_Uid_UID_2445_and_iCalUid`) we can just go back to what we had.

The main issue here is that if we just enqueue the removal of the extended property when there's an uid, it tries to remove it even when there's no `iCalUid`, so the query just fails. And the easiest way I have been able to find is by holding the properties available (in `extendedProperties`) and then checking whether it's present or not.

I'm using a `MutableList<String>`, but if you feel that it would be better to use a more "synchronized" method, for example with
```kotlin
@Volatile
var extendedProperties: List<String> = emptyList()
```
I think it may be useful.

**Comment by ArnyminerZ** [comment]
Looks good 👍🏼

**Review by rfc2822 — APPROVED** [review]

**Pull request merged**
